docs/architecture/adr-056-proving-amnesia-attacks.md

```diff
@@ -0,0 +1,120 @@
+
+### Amnesia Attack
+
+The currently only known form of global evidence stems from [flip flopping](https://github.com/tendermint/spec/blob/master/spec/consensus/light-client/accountability.md#flip-flopping) attacks. The schematic below explains one scenario where an amnesia attack, a form of flip flopping, can occur such that two sets of honest nodes, C1 and C2, commit different blocks.
+
+![](../imgs/tm-amnesia-attack.png)
+
+1. C1 and F send PREVOTE messages for block A.
+2. C1 sends PRECOMMIT for round 1 for block A.
+3. A new round is started, C2 and F send PREVOTE messages for a different block B.
+4. C2 and F then send PRECOMMIT messages for block B.
+5. F breaks the lock and goes back and sends PRECOMMIT messages in round 1 for block A.
+
+
+This creates a fork on the main chain.  Back to the past, another form of flip flopping, creates a light fork (capable of fooling those not involved in consensus), in a similar way, with F taking the precommits from C1 and forging a commit from them.
+
+## Decision
+
+As the distinction between these two attacks (amnesia and back to the past) can only be distinguished by confirming with all validators (to see if it is a full fork or a light fork), for the purpose of simplicity, these attacks will be treated as the same.
+
+Currently, the evidence reactor is used to simply broadcast and store evidence. Instead of perhaps creating a new reactor for the specific task of verifying these attacks, the current evidence reactor will be extended.
+
+The process begins with a light client receiving conflicting headers (in the future this could also be a full node during fast sync), which it sends to a full node to analyse. As part of [evidence handling](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-047-handling-evidence-from-light-client.md), this could be deduced into potential amnesia evidence
+
+```golang
+type PotentialAmnesiaEvidence struct {
+	V1 []*types.Vote
+	V2 []*types.Vote
+
+	timestamp time.Time
+}
+```
+
+*NOTE: Unlike prior evidence types, `PotentialAmnesiaEvidence` and `AmnesiaEvidence` are processed as a batch instead
+ of individually. This will require changes to much of the API.*
+
+ *NOTE: `PotentialAmnesiaEvidence` could be constructed for when 1/3 or less vote in two different rounds but as it is not currently detected nor can it cause a fork, it will be ignored.*
+
+The evidence should contain the precommit votes for the intersection of validators that voted for both rounds. The votes should be all valid and the height and time that the infringement was made should be within:
+
+`Trusting period - Amnesia trial period - Gossip safety margin`
+
+where `Amnesia trial period` is a configurable duration defaulted at 1 day and `Gossip safety margin` is the average time for all validators in a network to receive a message (an estimation based on the size of the network).
```

Comment:
also, we probably shouldn't add too many things to consensus params. can amnesia trial perid include gossip safety margin?

Reply:
won't we want to use gossip safety margin for other types of evidence (i.e. I invented it as a way of making sure that evidence is still valid when it finally is proposed because there will be a time gap in that)

Reply:
If we assume that EvidenceMaxAge incorporates this time and is therefore not equal to the Trust Period then I can just remove this